### ISSUE TITLE
Ballot Comment WD03 - Profile Specificity/Breadth

### ISSUE THREAD

**Issue by sparrell** [oasis-tcs/openc2-apsc-stateless-packet-filter#17]
Until we have many more actuator profiles, I think it is hard to determine how broad a profile should be. I believe this profile as constituted is really for two different functions and, as such, we are overburdening implementations with complexity and adding unnecessary complexity.
SLPF covers one major class of implementations including controls in many large cloud providers (e.g. AWS, Azure, ...) which implement stateless packet filters for ip-connection targets. Another use cases was proposed for certain hardware vendors which produce statues packet filters on ip-addresses only. Both of these are combined in this specification and I believe they should be separate profiles. As we develop more profiles we hopefully will develop guidelines on when to include-in-one-profile and when to split-into-several. Since any given security device is allowed to implement more than one profile, I think it is better to have smaller, more granular, profiles rather than combine them. My criteria would be to ask if there is a use case for a quick, painless switch betweeen the two choices. If the answer is yes (eg switching from AWS SLPF to Azure SLPF) then they should be in the same profile. But if the commands are incompatible (eg switching from a device with allow ip-connections to allow ip-address) then I would consider them different profile. It is possible to map an allow-ip-address (eg. Allow src-ip=1.1.1.1) to an allow-ip-connection (eg.  Allow src-ip=1.1.1.1, src-port-any, ...). However it is impossible to map most ip-connections (eg allow src-ip=1.1.1.1, Src-port=80) to ip-address without losing security controls and hence increasing security risk.

Because we combing ip-addr SLPF with ip-con SLFP it adds complexity in understanding (the option to one or the other for consumers) and it adds implementation burden for producers (eg a cloud orchestrator that will never encounter the fairly rare unique hardware will need to implement the capability for ip-addr which adds code, adds complexity, and IMHO adds security risk because of the added code which will not be used). Note my ‘fairly rare’ is a subjective opinion based on my career and I admit I only saw what I saw as a security architect in a large ISP. I am not against supporting ip-only functionality if someone needs it - I just don't think it should burden a very common, and growing functionality (ie in AWS, Azure, etc clouds).

In the interest of moving forward, I will vote yes (on WD03, 9/30/18) but I will continue to bring this up for reconsideration (eg at CSDPR01 12/12/2018) until we have many more profiles and have established more authoritatively how to define the breadth of a profile. And then it will either be accepted or a dead issue based on the criteria we develop.

**Comment by jmbrule** [comment]
The Actuator Profile SC and the participants at the TC discussed this matter at great length circa July 2018 (to include representatives from orchestration companies).  Your objection is noted but the general consensus of the balance of the members is with the current scope.  Will reopen as the TC gains experience.

**Comment by sparrell** [comment]
CSPRD01 comment https://www.oasis-open.org/apps/org/workgroup/openc2/email/archives/201812/msg00038.html reiterates this issue as noted in the CSPRDO1 CRM

**Comment by sparrell** [comment]
A straw poll on this issue is at https://wiki.oasis-open.org/openc2/ProfileSpecificityStrawPoll. Please add your opinions here and pick a choice on the straw poll (even if it's the 'either works for me' choice).

**Comment by TobyConsidine** [comment]
We faced similar issues in OBIX 1.1. 
OBIX 1.0, written when XML was the hottest trend in specifications did not distinguish between message patterns, i.e., REST and SOAP. The XML Schema Description (XSD) specification was not yet around, so we wrote up the XML used in prose, with examples. The result was, in many ways, a hot mess.
- Compatibility came to be defined as "interacts with the predominate vendor in my market"
- The US [mostly] used REST, Asia and Russia used SOAP, Europe used a mix.
- The simplest devices could interoperate, because they never came near the edge; complex systems interacted poorly.
We had addressed an complex interoperability challenge by creating a new non-interoperating spec(https://xkcd.com/927/). Even so, OBIX became the predominate interface for middleware in the buildings/IOT space.

This required OBIX 1.1, which was almost exclusively an interoperability release.

a) With immense effort, we created XSD that was not incompatible with the prose. A lot of effort would have been saved with some formalism up front, because the quirks were the hardest to get into XSD.
b) we could then test all the examples in the spec against the XSD. (see JADN discussions, ongoing)

This gave us a document that was finally consistent, and enabled us to build toward interoperabilty between US and Asian products.

C) We created a document "Common Encodings for OBIX"  (XML including EXI, Binary, JSON). 
C.1) For XML, we had dropped to under a dozen raw "types" in the XSD, and all other types were derived from those types. (Part of the core XSD). We said "see XSD, examples". 
C.2) For Binary and JSON, we defined how to represent each of the types.

D) We created a "Common Bindings for OBIX: REST". This essentially documented the choices of the primary North American vendor, with some allowances for fixing some first draft errors.
E) We created a "Common Bindings for OBIX: SOAP". This essentially documented the choices of the Asian and German vendors, with some effort at alignment.

When we were nearly complete, we got a reach-out from the Smart Television Alliance which was planning to use OBIX as part of the common platform for Television Apps. Because of how OBIX was now layered:

F) Members of the Alliance joined OASIS to create "Bindings for OBIX: Web Socket" 

They were able to create this specification in little time. The biggest delay was the interminable debates about how if interacting with your TV from a smart phone, and wi-fi went away, how to re-establish the session over carrier network.  

The STA work then ended up being simply "OBIX encoded in JSON and bound to Web Socket." This made it straight forward to get from Smart Television Apps to Thermostats (that already understood OBIX) and to irrigation/sprinkler systems, and to home automation... A bridge merely had to translate between encodings and change the binding. This worked because of careful language about "Binding Specifications that claim conformance to OBIX..." and "Encoding specifications that claim conformance to OBIX..."

Our work here is a bit more difficult because we are adding the notion of profiles that potentially extend the language. We are using JSON (in JADN) to specify the common language elements rather than XSD. This means that JADN must be rock-solid by the time we are done. 

We are overloading the profiles by putting sub-model, binding, and encoding all in one place. That is what the meat of this thread is about.

**Comment by alevere** [comment]
I think a common use case will simply be to block an IP address from a threat feed. One challenge is that may actually take two rules to implement in some packet filters. But, there are a myriad of different packet filters, and some do not need to create multiple rules to do this, or even use "rules" in the typical sense at all. I do not know that breaking out profiles solves this. Am I against a specify AWS/Azure/GCP profile? No, but I still think the first statement here is valid.

**Comment by sparrell** [comment]
Part of the issue is semantics. It is not a AWS/Azure use case issue. That was an example.

All the stateless filters I personnally have run across (of which the cloud ones qualify so I used as examples I care about) are 5-tuple filters and that is what I proposed when I made the original input to have a SLPF profile so we walk first with stateless and later address stateful packet filters (which are the vast majority of firewalls). Other members have expressed the need for 2-tuple (ip src/dst address) stateless packet filters due to their use cases with particular high-speed packet filters. 
IMHO a lot of our complication in drafting the SLPF spec came from the need to handle both in one profile - and they are not cross-compatible. A 5-tuple filter may be used as a 2-tuple filter (put 'any' in the other 3 fields) but a 2-tuple filter can not be used as a 5-tuple filter. This results in more options and "one of either x or y" working in the spec. Existing implementations (the clouds with 5-tuple and some specialized high-speed hardware filters for the 2-tuple) also are not interchangeable. Both are multi-vendor interchangeable within their domains of 5&2-tuple respectively, but none work across domains (in particular 5-tuple vendor 1 can not be replaced with 2-tuple vendor 2).. 

This original issue came from my belief that we that these are two different types of SLPF and should have separate profiles for these two cases.  But we don't and I'll live with that for now. I still believe in a few years when we have more experience with multiple profiles that we may revisit. But for now I think the spec works as is so I vote to close this issue.
